alchemy/src/esbuild/bundle.ts

```diff
@@ -136,70 +161,181 @@ export const Bundle = Resource(
-    if (outputFile === undefined && bundlePath === undefined) {
-      throw new Error("Failed to create bundle");
+      }
+      // If path was determined by outfile, use that as the primary path identifier
+      // otherwise esbuild generates a path relative to entrypoint in outdir for outputFiles
+      if (props.outfile) {
+        mainBundlePath = path.resolve(props.outfile);
+      } else if (props.outdir && mainBundlePath) {
+         // If outdir is specified, the path in outputFiles is often <outdir>/<entrypoint_basename>.js
+         // We might want to ensure this path is correctly relative or absolute as expected.
+         // For now, let's assume the path from esbuild is what we want to expose.
+      }
+
+
+    } else if (result.metafile && result.metafile.outputs && esbuildWriteBehavior) {
+      // esbuild was run with write: true, files are on disk, use metafile
+      for (const [outputPath, outputMeta] of Object.entries(result.metafile.outputs)) {
+        // The entryPoint in outputMeta is relative to the current working directory if `absWorkingDir` is not set,
+        // or relative to `absWorkingDir` if it is.
+        // We need to compare it to props.entryPoint, which might be relative or absolute.
+        const resolvedOutputEntryPoint = outputMeta.entryPoint ? path.resolve(props.absWorkingDir ?? process.cwd(), outputMeta.entryPoint) : undefined;
+        const resolvedPropsEntryPoint = path.resolve(props.absWorkingDir ?? process.cwd(), props.entryPoint);
+
+        if (outputMeta.entryPoint && resolvedOutputEntryPoint === resolvedPropsEntryPoint) {
+          if (outputPath.endsWith(".map")) {
+            // This is a source map associated with our main entry point
+            // but esbuild usually puts the map for the main output, not for the entrypoint itself if it's an input.
+            // This logic might need refinement if there are multiple entry points or complex output structures.
+            // For a single entry point, the .map file will be sibling to the output JS.
+          } else {
+            mainBundlePath = path.resolve(outputPath);
+            mainBundleContent = await fs.readFile(mainBundlePath, "utf-8");
+            mainBundleHash = crypto.createHash("sha256").update(mainBundleContent).digest("hex");
+            // Check for an associated source map file
+            const potentialMapPath = mainBundlePath + ".map";
+            if (result.metafile.outputs[potentialMapPath.substring(path.resolve(props.absWorkingDir ?? process.cwd()).length +1 )] || result.metafile.outputs[potentialMapPath.substring(process.cwd().length +1)]) {
+               // Need to check both relative to absWorkingDir and cwd for robustness if metafile paths are inconsistent
+              const mapKey = Object.keys(result.metafile.outputs).find(p => path.resolve(props.absWorkingDir ?? process.cwd(), p) === potentialMapPath);
+              if (mapKey) {
+                sourceMapOutput = {
```

Comment:
This looks a little out of control? Can we take a step back try and simplify? 

Source maps are either inline or separate files.

We need to support `bundle` and `noBundle` use-cases.

Maybe it's best to change bundling to generate a directory (or even a virtual in memory only if esbuild supports it) and then generalize the way we upload all of those files. 

Most of that is established already with `NoBundleResult`.

I am not sure what all this extra hashing is for